File: chad/src/main/java/org/woahoverflow/chad/commands/developer/ModifyPresence.java
package org.woahoverflow.chad.commands.developer;

import org.jetbrains.annotations.NotNull;
import org.woahoverflow.chad.core.ChadInstance;
import org.woahoverflow.chad.core.ChadVar;
import org.woahoverflow.chad.framework.handle.GuildHandler;
import org.woahoverflow.chad.framework.handle.MessageHandler;
import org.woahoverflow.chad.framework.obj.Command;
import org.woahoverflow.chad.framework.obj.Guild;
import sx.blah.discord.handle.impl.events.guild.channel.message.MessageEvent;
import sx.blah.discord.handle.obj.ActivityType;
import sx.blah.discord.handle.obj.StatusType;

import java.util.HashMap;
import java.util.List;
import java.util.stream.Collectors;

/**
 * Modify Chad's discord presence
 *
 * @author sho, codebasepw
 */
public class ModifyPresence implements Command.Class {
    @Override
    public final Runnable run(@NotNull MessageEvent e, @NotNull List<String> args) {
        return() -> {
            String prefix = (String) GuildHandler.getGuild(e.getGuild().getLongID()).getObject(Guild.DataType.PREFIX);

            // Checks if there's no arguments
            if (args.isEmpty()) {
                new MessageHandler(e.getChannel(), e.getAuthor()).sendPresetError(MessageHandler.Messages.INVALID_ARGUMENTS, prefix +
                        "modpresence **new presence**");
                return;
            }

            // The switch statement changes the message
            String message;
            switch (args.get(0)) {
                default:
                    // Builds the message
                    String formattedMessage = args.stream().map(str -> str + ' ').collect(Collectors.joining());

                    // Changes the presence
                    ChadInstance.cli.changePresence(ChadVar.statusType, ActivityType.PLAYING, formattedMessage.trim());

                    // Updates the message
                    message = "Changed presence to `" + formattedMessage.trim() + "`.";

                    // Updates the ChadVar variable
                    ChadVar.currentStatus = formattedMessage.trim();

                    break;
                case "rotate":
                    // Enables presence rotation
                    ChadVar.rotatePresence = true;

                    // Updates the message
                    message = "Enabled presence rotation.";
                    break;
                case "static":
                    // Disables presence rotation
                    ChadVar.rotatePresence = false;

                    // Updates the message
                    message = "Disabled presence rotation.";
                    break;
                case "add":
                    // Removes the option argument
                    args.remove(0);

                    // Builds the presence string
                    String string = args.stream().map(str -> str + ' ').collect(Collectors.joining());

                    // Adds the message into the rotation pool.
                    ChadVar.presenceRotation.add(string.trim());

                    // Updates the message
                    message = "Added `" + string.trim() + "` to rotation";
                    break;
                case "view":
                    StringBuilder stringBuilder = new StringBuilder();

                    for (String s : ChadVar.presenceRotation) {
                        stringBuilder.append('`').append(s).append("`, ");
                    }

                    message = stringBuilder.toString().substring(0, stringBuilder.toString().length() - 2);
                    break;
                case "status":
                    // Removes the option argument
                    args.remove(0);

                    if (args.get(0).equalsIgnoreCase("idle")) {
                        // Updates the ChadVar to IDLE
                        ChadVar.statusType = StatusType.IDLE;

                        // Changes the presence
                        e.getClient().changePresence(StatusType.IDLE, ActivityType.PLAYING, ChadVar.currentStatus);

                        // Updates the message
                        message = "Changed status type to `Idle`";
                        break;
                    }

                    if (args.get(0).equalsIgnoreCase("online")) {
                        // Updates the ChadVar to Online
                        ChadVar.statusType = StatusType.ONLINE;

                        // Changes the presence
                        e.getClient().changePresence(StatusType.ONLINE, ActivityType.PLAYING, ChadVar.currentStatus);

                        // Updates the message
                        message = "Changed status type to `Online`";
                        break;
                    }

                    if (args.get(0).equalsIgnoreCase("offline")) {
                        // Updates the ChadVar to Offline
                        ChadVar.statusType = StatusType.OFFLINE;

                        // Changes the presence
                        e.getClient().changePresence(StatusType.INVISIBLE, ActivityType.PLAYING, ChadVar.currentStatus);

                        // Updates the message
                        message = "Changed status type to `Offline`";
                        break;
                    }

                    if (args.get(0).equalsIgnoreCase("dnd")) {
                        // Updates the ChadVar to Do Not Disturb
                        ChadVar.statusType = StatusType.DND;

                        // Changes the presence
                        e.getClient().changePresence(StatusType.DND, ActivityType.PLAYING, ChadVar.currentStatus);

                        // Updates the message
                        message = "Changed status type to `Do Not Disturb`";
                        break;
                    }
                    new MessageHandler(e.getChannel(), e.getAuthor()).sendPresetError(MessageHandler.Messages.INVALID_ARGUMENTS, "modpresence status **dnd/offline/online/idle**");
                    return;
            }

            // Sends the message
            new MessageHandler(e.getChannel(), e.getAuthor()).sendMessage(message);
        };
    }

    @Override
    public final Runnable help(@NotNull MessageEvent e) {
        HashMap<String, String> st = new HashMap<>();
        st.put("modpresence <string>", "Changes the bots rich presence message.");
        st.put("modpresence status <dnd/offline/online/idle>", "Changes the bots status.");
        st.put("modpresence <static/rotate>", "Disables or enables the bot's presence rotation.");
        return Command.helpCommand(st, "Modify Presence", e);
    }
}

File: chad/src/main/java/org/woahoverflow/chad/commands/developer/ModifyPresence.kt
package org.woahoverflow.chad.commands.developer

import org.woahoverflow.chad.core.ChadInstance
import org.woahoverflow.chad.core.ChadVar
import org.woahoverflow.chad.framework.handle.GuildHandler
import org.woahoverflow.chad.framework.handle.MessageHandler
import org.woahoverflow.chad.framework.obj.Command
import org.woahoverflow.chad.framework.obj.Guild
import sx.blah.discord.handle.impl.events.guild.channel.message.MessageEvent
import sx.blah.discord.handle.obj.ActivityType
import sx.blah.discord.handle.obj.StatusType
import java.util.*

/**
 * Modify Chad's discord presence
 *
 * @author sho, codebasepw
 */
class ModifyPresence : Command.Class {
    override fun run(e: MessageEvent, args: MutableList<String>): Runnable {
        return Runnable {
            val prefix = GuildHandler.getGuild(e.guild.longID).getObject(Guild.DataType.PREFIX) as String

            // Checks if there's no arguments
            if (args.isEmpty()) {
                MessageHandler(e.channel, e.author).sendPresetError(MessageHandler.Messages.INVALID_ARGUMENTS, prefix + "modpresence [rotate/static/add/view/status/activity/new presence]")
                return@Runnable
            }

            when (args[0]) {
                "rotate" -> {
                    ChadVar.rotatePresence = true
                    MessageHandler(e.channel, e.author).sendMessage("Enabled presence rotation.")
                }
                "static" -> {
                    ChadVar.rotatePresence = false
                    MessageHandler(e.channel, e.author).sendMessage("Disabled presence rotation.")
                }

                "add" -> {
                    // Removes the option argument
                    args.removeAt(0)

                    val sb = StringBuilder()
                    for (arg in args) sb.append("$arg ")

                    ChadVar.presenceRotation.add(sb.toString().trim { it <= ' ' })
                    MessageHandler(e.channel, e.author).sendMessage("Added `" + sb.toString().trim { it <= ' ' } + "` to rotation")
                }

                "view" -> {
                    val stringBuilder = StringBuilder()

                    for (s in ChadVar.presenceRotation) stringBuilder.append("`$s`, ")

                    MessageHandler(e.channel, e.author).sendMessage(stringBuilder.toString().substring(0, stringBuilder.toString().length - 2))
                }

                "status" -> {
                    // Removes the option argument
                    args.removeAt(0)

                    ChadVar.statusType = when (args[0].toLowerCase()) {
                        "idle" -> StatusType.IDLE
                        "online" -> StatusType.ONLINE
                        "offline" -> StatusType.INVISIBLE
                        "dnd" -> StatusType.DND
                        else -> {
                            MessageHandler(e.channel, e.author).sendPresetError(MessageHandler.Messages.INVALID_ARGUMENTS, "modpresence status [dnd/offline/online/idle]")
                            return@Runnable
                        }
                    }

                    e.client.changePresence(ChadVar.statusType, ChadVar.activityType, ChadVar.currentStatus)
                    MessageHandler(e.channel, e.author).sendMessage("Changed status type to `${ChadVar.statusType.name}`")
                    return@Runnable
                }

                "activity" -> {
                    // Removes the option argument
                    args.removeAt(0)

                    ChadVar.activityType = when (args[0].toLowerCase()) {
                        "playing" -> ActivityType.PLAYING
                        "listening" -> ActivityType.LISTENING
                        "watching" -> ActivityType.WATCHING
                        else -> {
                            MessageHandler(e.channel, e.author).sendPresetError(MessageHandler.Messages.INVALID_ARGUMENTS, "modpresence activity [playing/listening/watching]")
                            return@Runnable
                        }
                    }

                    e.client.changePresence(ChadVar.statusType, ChadVar.activityType, ChadVar.currentStatus)
                    MessageHandler(e.channel, e.author).sendMessage("Changed activity type to `${ChadVar.activityType.name}`")
                    return@Runnable
                }

                else -> {
                    val sb = StringBuilder()
                    for (arg in args) sb.append("$arg ")

                    ChadVar.currentStatus = sb.toString().trim { it <= ' ' }
                    ChadInstance.cli.changePresence(ChadVar.statusType, ChadVar.activityType, ChadVar.currentStatus)
                    MessageHandler(e.channel, e.author).sendMessage("Changed presence to `${ChadVar.currentStatus}`.")
                }
            }
        }
    }

    override fun help(e: MessageEvent): Runnable {
        val st = HashMap<String, String>()
        st["modpresence activity [playing/listening/watching]]"] = "Changes the bot's activity."
        st["modpresence [string]"] = "Changes the bots rich presence message."
        st["modpresence status [dnd/offline/online/idle]"] = "Changes the bots status."
        st["modpresence [static/rotate]"] = "Disables or enables the bot's presence rotation."
        return Command.helpCommand(st, "Modify Presence", e)
    }
}
